### DISCUSSION TITLE
Future of Power Apps Component Framework (PCF), Platform Libraries and PCF Tooling

### DISCUSSION THREAD

**Discussion by hajekj** [microsoft/powerplatform-build-tools#679] [General]
This is a general feedback thread for PCF and our endeavors with it.

## Platform Libraries
[Reference](https://learn.microsoft.com/en-us/power-apps/developer/component-framework/react-controls-platform-libraries)

Platform Libraries have been in preview for over a year. What is the roadmap? When is it [going to GA](https://powerusers.microsoft.com/t5/Power-Apps-Pro-Dev-ISV/React-controls-and-platform-libraries-GA-date/m-p/1872464)?

### Current issues
1. Dependency version mismatches
    * You specify that you want React@16.8.6, but you end up with React@17.0.2 provided by Power Apps, [which is not backwards compatible](https://developer.microsoft.com/en-us/fluentui#/controls/web/callout) just out of box and requires changes. `pcf-start` ships with React@16.8.6, so technically, you are testing locally against a completely different dependency.
    * The same issue is with @fluentui/react packages. The team makes serious breaking changes from version to version, so it is simply not possible to say I want @fluentui/react@8.29.0 and receive something else (this is not the case at the moment however, just a future thing).
2. Not enough versions to choose from
    * For example, the @fluentui/react package is provided as v8.29.0 version, but there is already v8.113.1 which has a lot of fixes etc. When will these be available? Or simply let us bring over our own platform libraries, so we can maintain dependencies.
3. Standard Controls support
    * Migrating the existing controls to the `virtual` mode is PiTA, I would prefer if we could use platform libraries in Standard Controls, and not just react and fluent. There are other libraries which would come in really handy and make the overall size of the controls much smaller and loading much faster.
    * _Note:_ We are already using it, but doubt that it is supported in the way we are doing it.
5. Custom Platform Libraries
    * Let us bring our own platform libraries (or whatever custom libraries) we want. It will improve speed, sizes and performance.
    * **UPDATE:** If you follow my very [old blog post](https://blog.thenetw.org/2019/06/06/getting-react-and-office-ui-fabric-work-with-powerapps-component-framework/) you can start bringing your custom libraries and share them between PCF, but don't rely on the versions too much (if you need a specific version, bring it as a new library), since version dependencies are funky.
    * **UPDATE 27MAR2025**: This has been addressed by [Dependent Libraries](https://learn.microsoft.com/en-us/power-apps/developer/component-framework/dependent-libraries)

## PCF Tooling
I will try focus only on `pcf-scripts` as `pcf-start` would probably require a completely separate thread.

* Please stop breaking the versions. One version of `pcf-scripts` suddenly started requiring `.eslintrc.json` to be present and builds started crashing when updated. Unless the tools get released as a major version with breaking changes, I don't think that this should happen. Make it optional. I am a big fan of eslint and enforcing code-style and other rules, but it just can't be that disruptive, especially since generated projects include `"pcf-scripts": "^1"`  in `package.json`. 
* Make it open-source (both `pcf-scripts` and `pcf-start`). I bet community could easily and quickly add support for WebAPI calls to `pcf-start` and make controls more easily testable locally. Yes there are third party tools for it... But why have to use multiple tools, when we all can just focus effort into one...?
* Provide support for monorepo like Microsoft's [RushJS](https://rushjs.io/) or whatever is cool right now. We are really struggling with PCF build times, dependency size etc. There are some hardcoded paths in `pcf-scripts` and `pcf-start` which simply make it impossible to use.
* I am aware that I can fork your existing `pcf-scripts` ["compiled" code](https://www.npmjs.com/package/pcf-scripts?activeTab=code) and build off from there, but once again - I think it would benefit the community much more if we all shared the effort.

It's a lot of questions, but I believe these things are really crucial for this.

**Comment by MaceWindu1** [comment]
+1, please make it open-source

**Comment by szmarci** [comment]
Yes, please. Make it open-source.

**Comment by petrochuk** [comment]
Thank you for this detailed feedback. I want to provide some context on the scope of this repository: powerplatform-build-tools primarily contains the build automation tooling - specifically the build tasks, not the broader platform-side runtime components, library management, or dependency resolution systems. Many of the concerns raised around React/FluentUI versioning, platform libraries, and runtime behavior involve platform components beyond what this repository covers, so we're unable to provide a comprehensive roadmap for the entire PCF ecosystem from here. The Power Platform developer experience continues to evolve across multiple authoring approaches, and our investments adapt based on platform priorities and developer needs across that spectrum. We appreciate the constructive feedback and will continue to monitor community input as we evaluate future improvements to the tooling in this repository.
